blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitCloneReadSaveRequest.java

```diff
@@ -0,0 +1,164 @@
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import io.jenkins.blueocean.service.embedded.util.UserSSHKeyManager;
+import jenkins.model.Jenkins;
+import jenkins.plugins.git.GitSCMSource;
+import jenkins.plugins.git.traits.GitToolSCMSourceTrait;
+import jenkins.scm.api.trait.SCMSourceTrait;
+import org.apache.commons.io.FileUtils;
+import org.eclipse.jgit.transport.URIish;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.jenkinsci.plugins.gitclient.GitClient;
+
+/**
+ * Basic functional read/save using a clone of the remote
+ * @author kzantow
+ */
+class GitCloneReadSaveRequest extends GitReadSaveRequest {
+    private final GitTool gitTool;
+
+    private GitClient git;
+    private File repositoryPath;
+
+    public GitCloneReadSaveRequest(GitSCMSource gitSource, String branch, String commitMessage, String sourceBranch, String filePath, byte[] contents) {
+        super(gitSource, branch, commitMessage, sourceBranch, filePath, contents);
+
+        GitTool.DescriptorImpl toolDesc = Jenkins.getInstance().getDescriptorByType(GitTool.DescriptorImpl.class);
+        @SuppressWarnings("deprecation")
+        GitTool foundGitTool = toolDesc.getInstallation(gitSource.getGitTool());
+        for (SCMSourceTrait trait : gitSource.getTraits()) {
+            if (trait instanceof GitToolSCMSourceTrait) {
+                foundGitTool = toolDesc.getInstallation(((GitToolSCMSourceTrait) trait).getGitTool());
+            }
+        }
+        if (foundGitTool == null) {
+            foundGitTool = GitTool.getDefaultInstallation();
+        }
+
+        this.gitTool = foundGitTool;
+    }
+
+    private void cloneRepo() throws InterruptedException, IOException {
+        repositoryPath = Files.createTempDirectory("git").toFile();
+
+        EnvVars environment = new EnvVars();
+        TaskListener taskListener = new LogTaskListener(Logger.getAnonymousLogger(), Level.ALL);
+        String gitExe = gitTool.getGitExe();
+        git = Git.with(taskListener, environment)
+                .in(repositoryPath)
+                .using(gitExe)
+                .getClient();
+
+        if (GitUtils.isSshUrl(gitSource.getRemote())) {
+            // Set user ssh credentials
+            User user = User.current();
+            if (user == null) {
+                throw new ServiceException.UnauthorizedException("Not authenticated");
+            }
+            BasicSSHUserPrivateKey privateKey = UserSSHKeyManager.getOrCreate(user);
+            git.addCredentials(gitSource.getRemote(), privateKey);
+        }
+
+        git.clone(gitSource.getRemote(), "origin", true, null);
+
+        log.fine("Repository; " + gitSource.getRemote() + " cloned to: " + repositoryPath.getCanonicalPath());
+    }
+
+    @Override
+    byte[] read() throws IOException {
+        try {
+            cloneRepo();
```

Comment:
`cloneRepo()` creates and sets `git` and `repositoryPath` non-final class instance variable. Its cleaner for `cloneRepo()` to return GitClient instance and remove these class instance variable. 

```java
GitClient git = cloneRepo();
...
```

In fact GitClient creation along with ssh credential search should be moved to a utility as I see similar code gets repeated at different places.